woocommerce/woocommerce
# Performance - Cache full product objects

### All Submissions:






### Changes proposed in this Pull Request:

I was doing performance profiling around the Shop page and my profiling placed the product constructors near the top areas by time. This makes sense since loading products to display is basically all the Shop page does, so that’s not super interesting info on its own. The interesting part is that the profiling identified that there were many more requests to load data than there were products displayed on the Shop page. For example, we loaded product data 51 times on a Shop page with 16 products, and each product displayed in the Shop was loaded 3 times.

We do have plenty of caching around individual data pieces, and WordPress has caching around metadata. The main performance issue I found was that each time we load a product’s data 100+ functions run that set properties, sanitize data, apply filters, etc. We can improve performance quite a lot by not having to run these functions when fetching products we have already fetched.

When doing e.g. `get_post( $id )` multiple times for the same post, WordPress core [caches the full post objects](https://github.com/WordPress/WordPress/blob/7d17b53ff2882b09f7071bfa25a7d5972b3924a8/wp-includes/class-wp-post.php#L231-L255). A similar approach could help here, and in testing this approach improved performance on the Shop page dramatically.

I'm told something similar used to be in the product factory, but @franticpsyx noticed issues with it and it was removed. @franticpsyx, do you recall what the problem was in the past?

### Todo

- Product images on a lot of products disappear now for some reason. Other bugs may also be present. They should be investigated + fixed.
- Unit tests need to pass. Adding more/better clearing of the cached products should help with that.

### How to test the changes in this Pull Request:

These testing steps will not currently pass:
1. Verify product-based WooCommerce functionality (Shop, single products, admin, etc.) works correctly.
2. Verify unit tests pass.

### Other information:




### Changelog entry

> Performance - Cache full products to improve performance when loading the same product multiple times.
